state/cockroachdb/cockroachdb_access.go

```diff
@@ -411,54 +508,93 @@ func (p *cockroachDBAccess) Ping(ctx context.Context) error {
 	backoff := config.NewBackOff()
 
 	return retry.NotifyRecover(func() error {
-		err := p.db.PingContext(ctx)
+		err := c.db.PingContext(ctx)
 		if errors.Is(err, driver.ErrBadConn) {
 			return fmt.Errorf("error when attempting to establish connection with cockroachDB: %v", err)
 		}
 		return nil
 	}, backoff, func(err error, _ time.Duration) {
-		p.logger.Debugf("Could not establish connection with cockroachDB. Retrying...: %v", err)
+		c.logger.Debugf("Could not establish connection with cockroachDB. Retrying...: %v", err)
 	}, func() {
-		p.logger.Debug("Successfully established connection with cockroachDB after it previously failed")
+		c.logger.Debug("Successfully established connection with cockroachDB after it previously failed")
 	})
 }
 
 // Close implements io.Close.
-func (p *cockroachDBAccess) Close() error {
-	if p.db != nil {
-		return p.db.Close()
+func (c *CockroachDBAccess) Close() error {
+	if c.closed.CompareAndSwap(false, true) {
+		close(c.closeCh)
+	}
+	defer c.wg.Wait()
+
+	if c.db != nil {
+		return c.db.Close()
 	}
 
 	return nil
 }
 
-func (p *cockroachDBAccess) ensureStateTable(stateTableName string) error {
-	exists, err := tableExists(p.db, stateTableName)
+func (c *CockroachDBAccess) ensureStateTable(ctx context.Context, stateTableName string) error {
+	exists, err := tableExists(ctx, c.db, stateTableName)
 	if err != nil {
 		return err
 	}
 
 	if !exists {
-		p.logger.Info("Creating CockroachDB state table")
-		createTable := fmt.Sprintf(`CREATE TABLE %s (
-									key text NOT NULL PRIMARY KEY,
-									value jsonb NOT NULL,
-									isbinary boolean NOT NULL,
-									etag INT,
-									insertdate TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT NOW(),
-									updatedate TIMESTAMP WITH TIME ZONE NULL);`, stateTableName)
-		_, err = p.db.Exec(createTable)
+		c.logger.Info("Creating CockroachDB state table")
+		_, err = c.db.ExecContext(ctx, fmt.Sprintf(`CREATE TABLE %s (
+  key text NOT NULL PRIMARY KEY,
+  value jsonb NOT NULL,
+  isbinary boolean NOT NULL,
+  etag INT,
+  insertdate TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT NOW(),
+  updatedate TIMESTAMP WITH TIME ZONE NULL,
+  expire_at TIMESTAMP WITH TIME ZONE NULL,
```

Comment:
Can you call this `expiredate` so it's the same as in the Postgres component? (all other columns are the same except etag which doesn't exist in postgres since we use the xid)

Reply:
Done.